### PR TITLE
Improve USB boot stability

### PR DESCRIPTION
**Pull request by jviotti**
- Check the `INIT_ERROR` flag when initializing `libusb`
- Don't wait infinitely for USB transfers
- Retry if a bulk transfer stalls

Change-type: patch

### CODE REVIEW

**Comment by jviotti** [comment]
Mostly back ported from https://github.com/balena-io/etcher/tree/1-4-line-firmware-backport/lib/sdk/adapters/usbboot

**Comment by jviotti** [comment]
The delays may look weird at the beginning, but based on what I've seen, sending the blobs to fast can cause USB stalls, and if that happens, we will have to wait until the timeout to then try again. If we add a small delay at the beginning, the chances of success at first try are much greater, which ends up being way faster than a whole timeout/retry cycle.

File: lib/index.ts [hunk 1]
```diff
@@ -396,6 +428,11 @@ export class UsbbootScanner extends EventEmitter {
 		super();
 		this.boundAttachDevice = this.attachDevice.bind(this);
 		this.boundDetachDevice = this.detachDevice.bind(this);
+
+		// This is an undocumented property
+		if ((<any>usb).INIT_ERROR as any) {
```
Comment (jviotti) [inline lib/index.ts]: @zvin @thundron Is there a better way to make TSLint happy here? `INIT_ERROR` is not on the `usb` types we're pulling in

File: lib/index.ts [hunk 2]
```diff
@@ -253,17 +259,43 @@ function* chunks(buffer: Buffer, size: number) {
...
+	} catch (error) {
+		if (error.errno === usb.LIBUSB_TRANSFER_STALL) {
+			debug('Transfer stall, retrying');
+			await transfer(endpoint, chunk, times - 1);
+		}
+
+		throw error;
+	}
+};
+
```
Comment (zvin) [inline lib/index.ts]: You can avoid recursion here
```typescript
const transfer = async (endpoint: usb.OutEndpoint, chunk: Buffer) => {
	endpoint.timeout = USB_BULK_TRANSFER_TIMEOUT_MS;
	for (let tries = 0; tries < 3; tries++) {
		if (tries > 0) {
			debug('Transfer stall, retrying');
		}
		try {
			await fromCallback(callback => {
				endpoint.transfer(chunk, callback);
			});
			return;
		} catch (error) {
			if (error.errno === usb.LIBUSB_TRANSFER_STALL) {
				continue;
			}
			throw error;
		}
	}
};
```

File: lib/index.ts [hunk 3]
```diff
@@ -110,6 +113,9 @@ const DEVICE_UNPLUG_TIMEOUT = 5000;
 // Delay (in ms) to wait when epRead throws an error
 const READ_ERROR_DELAY = 100;
 
+// Transfer delay
+const TRANSFER_DELAY_MS = 1000;
```
Comment (zvin) [inline lib/index.ts]: This code works well with no delay added.
This change makes `node-raspberrypi-usbboot` 33 seconds slower for each device.

File: lib/index.ts [hunk 4]
```diff
@@ -396,6 +428,11 @@ export class UsbbootScanner extends EventEmitter {
 		super();
 		this.boundAttachDevice = this.attachDevice.bind(this);
 		this.boundDetachDevice = this.detachDevice.bind(this);
+
+		// This is an undocumented property
+		if ((usb as any).INIT_ERROR as any) {
```
Comment (zvin) [inline lib/index.ts]: You can just add a `// @ts-ignore` comment on the preceding line:
```typescript
// @ts-ignore
if (usb.INIT_ERROR) {
```

**Pull request merged**